src/languageClientConsumer.ts

```diff
@@ -1,28 +1,66 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-import { window } from "vscode";
+import { ProgressLocation, window } from "vscode";
 import { LanguageClient } from "vscode-languageclient/node";
 
 export abstract class LanguageClientConsumer {
+    private static languageClientPromise?: Promise<LanguageClient>;
+    private static getLanguageClientResolve?: (value: LanguageClient) => void;
 
-    private _languageClient: LanguageClient | undefined;
+    // Implementations of this class must override this method to register their
+    // handlers, as its called whenever the client is restarted / replaced.
+    public abstract onLanguageClientSet(languageClient: LanguageClient): void;
 
-    public setLanguageClient(languageClient: LanguageClient): void {
-        this.languageClient = languageClient;
+    // This is called in the session manager when the client is started (so we
+    // can wait for that). It's what actually resolves the promise.
+    public static onLanguageClientStarted(languageClient: LanguageClient): void {
+        // It should have been created earlier, but if not, create and resolve it.
+        this.languageClientPromise ??= Promise.resolve(languageClient);
+        this.getLanguageClientResolve?.(languageClient);
     }
 
-    abstract dispose(): void;
+    // This is called in the session manager when the client exits so we can
+    // make a new promise.
+    public static onLanguageClientExited(): void {
+        this.languageClientPromise = undefined;
+        this.getLanguageClientResolve = undefined;
+    }
 
-    public get languageClient(): LanguageClient | undefined {
-        if (!this._languageClient) {
-            // TODO: Plumb through the logger.
-            void window.showInformationMessage("PowerShell extension has not finished starting up yet. Please try again in a few moments.");
-        }
-        return this._languageClient;
+    // We should have a promise as defined in resetLanguageClient, but if we
+    // don't, create it.
+    public static async getLanguageClient(): Promise<LanguageClient> {
```

Comment:
I feel like this would be more suited to a lazy getter than encapsulates all the other logic and returns a `Promise<LanguageClient>`. That way derived Consumers can just reference the property and the first one to do so triggers a language client startup e.g. `await this.languageClient` or `await this.languageClientPromise` (I prefer the former)

Reply:
That said this is totally fine too

Reply:
After further review I see this class is basically using languageClient as a singleton. If that's the case then why does everything have to inherit/consume from this class vs. it just be a standalone singleton class that has getLanguageClient as a static async method, and a onLanguageClientStarted EventEmitter (using either the vscode flavor or the nodeJS flavor) that clients can register to? That would enable them to be decoupled from this base class and be whatever they want (or have this functionality put in an interface)

Reply:
https://vscode-docs.readthedocs.io/en/stable/extensions/patterns-and-principles/#events
https://refactoring.guru/design-patterns/singleton/typescript/example